### PR TITLE
Changed smart contracts to be seperate token and contract

### PR DESCRIPTION
**Pull request by DavidGershony**

### CODE REVIEW

File: src/Blockcore.Indexer.Cirrus/Storage/Mongo/SmartContracts/ComputeSmartContractServiceWithSplitDocuments.cs [hunk 1]
```diff
@@ -13,55 +13,100 @@
...
+      ISmartContractTransactionsLookup<T> transactionsLookup)
    {
       this.logger = logger;
-      mongoDb = db;
+      this.mongoDb = mongoDb;
       this.logReaderFactory = logReaderFactory;
-      this.mongoDatabase = mongoDatabase;
       cirrusClient = (CirrusClient)clientFactory.Create(connection);
-      emptyContract = new T();
+      ;
```
Comment (dangershony) [inline src/Blockcore.Indexer.Cirrus/Storage/Mongo/SmartContracts/ComputeSmartContractServiceWithSplitDocuments.cs:35]: extra `;`

File: src/Blockcore.Indexer.Cirrus/Sync/SyncTasks/SmartContractSyncRunner.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+          Runner.GlobalState.ReorgMode ||
+          Runner.GlobalState.StoreTip == null ||
+          Runner.GlobalState.IndexMode)
+      {
+         return false;
+      }
+
+      logger.LogDebug($"Calling {nameof(storage.GetSmartContractsThatNeedsUpdatingAsync)}");
+
+      var smartContractAddresses = await storage.GetSmartContractsThatNeedsUpdatingAsync(supportedSmartContractTypes.Keys.ToArray());
```
Comment (dangershony) [inline src/Blockcore.Indexer.Cirrus/Sync/SyncTasks/SmartContractSyncRunner.cs:61]: don't you need to pas a block height here?

**Review by dangershony — COMMENTED** [review]
some minor comments but look fine

File: src/Blockcore.Indexer.Cirrus/Storage/Mongo/CirrusBlockRewindOperation.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      var smartContractsToBeDeleted = ((ICirrusMongoDb)storage).SmartContractTable.AsQueryable()
+         .Where(_ => _.LastProcessedBlockHeight >= blockIndex)
+         .ToList();
+
+      var rewindTasks = smartContractsToBeDeleted.Select(_ => DeleteTokensForSmartContractAddressAsync(_.ContractAddress));
+
+      await Task.WhenAll(rewindTasks);
+   }
+
+   Task DeleteTokensForSmartContractAddressAsync(string address)
```
[the remark this answers is not in the document]
Comment (DavidGershony) [reply src/Blockcore.Indexer.Cirrus/Storage/Mongo/CirrusBlockRewindOperation.cs:26]: Yes, if a contract has changes after the block rewind I just delete all contract data and the runner will pick that contract up and recreate all the data

File: src/Blockcore.Indexer.Cirrus/Storage/Mongo/SmartContracts/Dao/GetOrPropertyCallsLogReader.cs [hunk 4]
```diff
@@ -1,17 +1,20 @@
...
 
    public bool IsTransactionLogComplete(LogResponse[] logs) => true;
 
-   public void UpdateContractFromTransactionLog(CirrusContractTable contractTransaction,
-      DaoContractComputedTable computedTable)
-   { }
+   public WriteModel<DaoContractProposalTable>[] UpdateContractFromTransactionLog(CirrusContractTable contractTransaction,
+      DaoContractTable computedTable)
+   {
+      return null;
```
[the remark this answers is not in the document]
Comment (DavidGershony) [reply src/Blockcore.Indexer.Cirrus/Storage/Mongo/SmartContracts/Dao/GetOrPropertyCallsLogReader.cs:18]: We don't actually have access to the data without extracting the parameters from the call to the smart contract.
Might add that one later on if we need it but it's a big piece of work.

**Pull request merged**